File: 卡码/代码随想录/1/06-二叉树/Js/05-二叉树公共祖先/LeetCode_0236.js
//236. 二叉树的最近公共祖先
import TreeNode from "../TreeNode.js"

const lowestCommonAncestor = function(root, p, q) {
  if(!root||root==p||root==q) return root;
  const left=lowestCommonAncestor(root.left,p,q);
  const right=lowestCommonAncestor(root.right,p,q);
  if(left&&right) return root;
  else if(!left&&right) return right;
  else if(left&&!right) return left;
  else return null;
};

File: 卡码/代码随想录/1/06-二叉树/Go/05-二叉树公共祖先/LeetCode_0236.go
//236. 二叉树的最近公共祖先
package main

import (
	. "binaryTree/types"
)

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
  if root==nil||root==p||root==q {
		return root
	}
	left:=lowestCommonAncestor(root.Left,p,q)
	right:=lowestCommonAncestor(root.Right,p,q)
	if left!=nil&&right!=nil {
		return root
	}else if left==nil&&right!=nil {
		return right
	}else if left!=nil&&right==nil {
		return left
	}else{
		return nil
	}
}
